### PR TITLE
fix: use O(num_classes) memory in multiclass stat scores (fixes #3343)

### PR DESCRIPTION
**Pull request by jlaportebot**
## What does this PR do?

Fixes #3343

Replaces the O(num_classes**2) confusion-matrix computation in `_multiclass_stat_scores_update` with an O(num_classes) linear-memory approach. Previously, calling `MulticlassAccuracy(average="macro", num_classes=1_000_000)` would cause OOM. Now it works with constant memory.

### Changes

**`src/torchmetrics/functional/classification/stat_scores.py`**: Replace confusion-matrix-based TP/FP/FN/TN with direct bincount computation:
- `TP[c] = bincount(preds[target==preds], minlength=num_classes)`
- `FP[c] = bincount(preds) - TP[c]`
- `FN[c] = bincount(target) - TP[c]`
- `TN[c] = total - (TP[c] + FP[c] + FN[c])`

This is mathematically equivalent to the confusion matrix diagonal/row/column sums.

### Reproduction

```python
import torch
from torchmetrics.classification import MulticlassAccuracy

metric = MulticlassAccuracy(average="macro", num_classes=1_000_000)
metric(torch.tensor([1]), torch.tensor([1]))  # <-- OOM before, works now
```

### Tests

- Added `test_multiclass_stat_scores_linear_memory` parametrized over num_classes=[5, 50, 1M] and average=[micro, macro, weighted, None]
- Added `test_multiclass_accuracy_large_num_classes` for the 1M classes accuracy case
- All 241 existing multiclass stat score tests pass

### CODE REVIEW

File: tests/unittests/classification/test_stat_scores.py [hunk 1]
```diff
@@ -476,6 +476,46 @@ def test_multiclass_overflow():
...
 
+@pytest.mark.parametrize("num_classes", [5, 50, 1_000_000])
+@pytest.mark.parametrize("average", ["micro", "macro", "weighted", None])
+def test_multiclass_stat_scores_linear_memory(num_classes, average):
+    """Test that multiclass stat scores use O(num_classes) memory instead of O(num_classes**2).
+
+    Regression test for https://github.com/Lightning-AI/torchmetrics/issues/3343.
+
+    """
+    n = 500
```
Comment (Borda) [inline tests/unittests/classification/test_stat_scores.py:487]: let's have this as an arg  toowith a simple default = 500

**Review by Borda — APPROVED** [review]
